x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java

```diff
@@ -595,11 +596,23 @@ GetUserPrivilegesResponse buildUserPrivilegesResponseObject(Role userRole) {
     }
 
     static Set<String> resolveAuthorizedIndicesFromRole(Role role, RequestInfo requestInfo, Map<String, IndexAbstraction> lookup) {
-        Predicate<IndexAbstraction> predicate = role.allowedIndicesMatcher(requestInfo.getAction());
+        final Predicate<IndexAbstraction> predicate = role.allowedIndicesMatcher(requestInfo.getAction());
 
         // do not include data streams for actions that do not operate on data streams
         TransportRequest request = requestInfo.getRequest();
         final boolean includeDataStreams = (request instanceof IndicesRequest) && ((IndicesRequest) request).includeDataStreams();
+        if (predicate == Role.MATCH_ALL_INDICES_MATCHER) {
+            if (includeDataStreams) {
+                return lookup.values().stream()
+                    .map(IndexAbstraction::getName)
+                    .collect(Collectors.toUnmodifiableSet());
+            } else {
+                return lookup.values().stream()
+                    .filter(indexAbstraction -> indexAbstraction.getType() != IndexAbstraction.Type.DATA_STREAM)
+                    .map(IndexAbstraction::getName)
+                    .collect(Collectors.toUnmodifiableSet());
+            }
+        }
```

Comment:
If #78321 is merged, this change will only take effect if the logic in #78321 decides wildcard expansion must be performed. This is good because avoid wildcard expansion is still much more efficient than returning a large list of names here.

Reply:
After some testing, this change does not really offer much improvement over existing code since #77738 has already short-circuited the underlying predicate. I guess if the cluster has a large number of data streams and each data stream has a lot of back indices, this might help avoid double looping on the backing indices. But otherwise, it is not very useful. I am OK to remove this part of change if necessary.

Reply:
After some testing, this change does not really offer much improvement over existing code since #77738 has already short-circuited the underlying predicate. I guess if the cluster has a large number of data streams and each data stream has a lot of back indices, this might help avoid double looping on the backing indices. But otherwise, it is not very useful. I am OK to remove this part of change if necessary.

Reply:
I decided to remove this change.